### PR TITLE
Bug 1952611: [OCPCLOUD-1115] Get instance tags from infrastructure object

### PR DESCRIPTION
**Pull request by alexander-demichev**
Get instance tags from infrastructure objects. Tags that come from the infra object have precedence over providerSpec.

Related enhancements: https://github.com/openshift/enhancements/pull/706

### CODE REVIEW

File: pkg/actuators/machine/instances.go [hunk 1]
```diff
@@ -422,9 +427,27 @@ func buildTagList(machineName string, clusterID string, machineTags []awsprovide
...
+
+	for _, tag := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		mergedList = append(mergedList, awsproviderv1.TagSpecification{Name: tag.Key, Value: tag.Value})
+	}
+
+	mergedList = append(mergedList, machineSpecTags...)
+
+	return mergedList
+}
+
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:450]: Which of these takes precedence? We updated the agreement to a local override, so we need to make sure that Machine tags override the Infra tags

Please make sure this is tested somewhere and that a comment is added to the function comment so we note this in the future

File: pkg/actuators/machine/instances_test.go [hunk 2]
```diff
@@ -58,35 +59,128 @@ func TestRemoveDuplicatedTags(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			expected: []*ec2.Tag{
+				{Key: aws.String("good"), Value: aws.String("goodvalue")},
+				{Key: aws.String("kubernetes.io/cluster/clusterID"), Value: aws.String("owned")},
+				{Key: aws.String("Name"), Value: aws.String("machineName")},
+			},
+		},
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances_test.go:175]: Ok so we need to reverse this test based on the latest version of the enhancement that was agreed

File: pkg/actuators/machine/reconciler.go [hunk 3]
```diff
@@ -64,7 +67,14 @@ func (r *Reconciler) create() error {
 		return fmt.Errorf("failed to get user data: %w", err)
 	}
 
-	instance, err := launchInstance(r.machine, r.providerSpec, userData, r.awsClient)
+	infra := &configv1.Infrastructure{}
+	infraName := client.ObjectKey{Name: awsclient.GlobalInfrastuctureName}
+
+	if err := r.client.Get(r.Context, infraName, infra); err != nil {
+		return err
```
Comment (JoelSpeed) [inline pkg/actuators/machine/reconciler.go:74]: Please add some context to this error and add a log line please

Do we want to fail if there is no infra object? ie should we be accounting for a not found error at all?

**Review by JoelSpeed — COMMENTED** [review]
Looks good, need to fix the ordering of the tag precedence as I think the conversation in the proposal changed the ordering between this being drafted and now

File: pkg/actuators/machine/reconciler.go (hunk 3 above)
Comment (alexander-demichev) [reply pkg/actuators/machine/reconciler.go:74]: I think that infra object should always be present.

File: pkg/actuators/machine/instances_test.go (hunk 2 above)
Comment (alexander-demichev) [reply pkg/actuators/machine/instances_test.go:175]: thanks, I'll fix it in a minute

File: pkg/actuators/machine/instances.go [hunk 4]
```diff
@@ -422,9 +427,28 @@ func buildTagList(machineName string, clusterID string, machineTags []awsprovide
...
 		{Key: aws.String("Name"), Value: aws.String(machineName)},
 	}...)
+
 	return removeDuplicatedTags(rawTagList)
 }
 
+// mergeInfrastructureAndMachineSpecTags merge list of tags from machine provider spec and Infrastructure object platform spec.
+// Machine tags have precedence over Infrastructure
+func mergeInfrastructureAndMachineSpecTags(machineSpecTags []awsproviderv1.TagSpecification, infra *configv1.Infrastructure) []awsproviderv1.TagSpecification {
+	if infra == nil || infra.Status.PlatformStatus == nil || infra.Status.PlatformStatus.AWS == nil || infra.Status.PlatformStatus.AWS.ResourceTags == nil {
```
Comment (michaelgugino) [inline pkg/actuators/machine/instances.go:437]: Check for nil infra.Status here as well?  I don't think that status is always guaranteed to be present.

Comment (JoelSpeed) [reply pkg/actuators/machine/instances.go:437]: It's not a pointer so I don't think it's required https://github.com/openshift/api/blob/a3c8fa53e01bc4442b687d3aca1571692f7676ce/config/v1/types_infrastructure.go#L11-L22

File: pkg/actuators/machine/instances.go [hunk 5]
```diff
@@ -422,9 +427,28 @@ func buildTagList(machineName string, clusterID string, machineTags []awsprovide
...
+		return machineSpecTags
+	}
+
+	mergedList := []awsproviderv1.TagSpecification{}
+
+	for _, tag := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		mergedList = append(mergedList, awsproviderv1.TagSpecification{Name: tag.Key, Value: tag.Value})
+	}
+
+	machineSpecTags = append(machineSpecTags, mergedList...)
```
Comment (michaelgugino) [inline pkg/actuators/machine/instances.go:447]: This looks like we're allowing for duplicates in the list which is unsupported I believe.  Also, the enhancement specifies tags set on the machineset should override those in the infra.Status, so we need to build a map, first with the infra.Status tags, then apply the machineset, then dump to list.

**Review by michaelgugino — CHANGES_REQUESTED** [review]

File: pkg/actuators/machine/instances.go (hunk 5 above)
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:447]: Best not to modify the machineSpecTags list as that is a reference value, let's continue to use a fresh list, does this suggestion work?
```suggestion
	mergedList := []awsproviderv1.TagSpecification{machineSpecTags...}

	for _, tag := range infra.Status.PlatformStatus.AWS.ResourceTags {
		mergedList = append(mergedList, awsproviderv1.TagSpecification{Name: tag.Key, Value: tag.Value})
	}
```

Comment (JoelSpeed) [reply pkg/actuators/machine/instances.go:447]: The tags returned from this list are deduplicated once the name and cluster tags are added, look at approx line 431 of this file. Based on the unit testing, I think this is all working as expected

Comment (alexander-demichev) [reply pkg/actuators/machine/instances.go:447]: yes, tags are deduplicated later

Comment (alexander-demichev) [reply pkg/actuators/machine/instances.go]: that works too

File: pkg/actuators/machine/instances.go [hunk 6]
```diff
@@ -422,9 +427,29 @@ func buildTagList(machineName string, clusterID string, machineTags []awsprovide
...
+	mergedList = append(mergedList, machineSpecTags...)
+
+	for _, tag := range infra.Status.PlatformStatus.AWS.ResourceTags {
+		mergedList = append(mergedList, awsproviderv1.TagSpecification{Name: tag.Key, Value: tag.Value})
+	}
+
+	machineSpecTags = append(machineSpecTags, mergedList...)
+
+	return machineSpecTags
+}
```
Comment (JoelSpeed) [inline pkg/actuators/machine/instances.go:451]: We currently have this duplicated now
```suggestion
	mergedList := []awsproviderv1.TagSpecification{}
	mergedList = append(mergedList, machineSpecTags...)

	for _, tag := range infra.Status.PlatformStatus.AWS.ResourceTags {
		mergedList = append(mergedList, awsproviderv1.TagSpecification{Name: tag.Key, Value: tag.Value})
	}

	return mergedList
}
```

Comment (alexander-demichev) [reply pkg/actuators/machine/instances.go:451]: 🤦‍♂️

**Review by michaelgugino — APPROVED** [review]

**Comment by JoelSpeed** [comment]
/retitle [OCPCLOUD-1115] Get instance tags from infrastructure object

**Comment by JoelSpeed** [comment]
/retitle Bug 1952611: [OCPCLOUD-1115] Get instance tags from infrastructure object

**Pull request merged**